ui/core/src/model/calculations.ts

```diff
@@ -56,38 +60,52 @@ export const CALCULATIONS_CONFIG: Readonly<Record<CalculationType, CalculationCo
 
 export const DEFAULT_CALCULATION: CalculationType = 'Sum';
 
-function first(values: TimeSeriesValueTuple[]): number | undefined {
+type CalculationValue = number | null | undefined;
+
+function first(values: TimeSeriesValueTuple[]): CalculationValue {
   const tuple = values[0];
   return tuple === undefined ? undefined : getValue(tuple);
 }
 
-function last(values: TimeSeriesValueTuple[]): number | undefined {
+function last(values: TimeSeriesValueTuple[]): CalculationValue {
   if (values.length <= 0) return undefined;
 
   const tuple = values[values.length - 1];
   return tuple === undefined ? undefined : getValue(tuple);
 }
 
-function lastNumber(values: TimeSeriesValueTuple[]): number | undefined {
-  const tuple = findLast(values, (tuple) => isNaN(getValue(tuple)) === false);
+function firstNumber(values: TimeSeriesValueTuple[]): CalculationValue {
+  const tuple = values.find((tuple) => typeof getValue(tuple) === 'number');
+  return tuple === undefined ? undefined : getValue(tuple);
+}
+
+function lastNumber(values: TimeSeriesValueTuple[]): CalculationValue {
+  const tuple = findLast(values, (tuple) => typeof getValue(tuple) === 'number');
   return tuple === undefined ? undefined : getValue(tuple);
 }
 
-function mean(values: TimeSeriesValueTuple[]): number | undefined {
+function mean(values: TimeSeriesValueTuple[]): CalculationValue {
   if (values.length <= 0) return undefined;
-  return meanBy(values, getValue);
+  return lodashMean(getNonNullValues(values));
 }
 
-function sum(values: TimeSeriesValueTuple[]): number | undefined {
+function sum(values: TimeSeriesValueTuple[]): CalculationValue {
   if (values.length <= 0) return undefined;
-  return sumBy(values, getValue);
```

Comment:
Honestly, we could probably get rid of all the lodash calc functions in here. I'm going to wait until my next PR where I look at calculating multiple values at once for table legend to take a pass at doing that.

Reply:
Due to bundle size concerns or just that these are so simple they would be clearer w/o lodash? @saminzadeh and I have talked about this in other cases too and I know he prefers using battle tested libraries for this type of thing but there are definitely tradeoffs

Reply:
From a perf perspective, it'd be a bummer to iterate a long array n times where n is the number of calculations we need to do (so we can use each of those lodash functions) vs. having a single loop where we can do all the calculations at once. These are all pretty simple math, so I don't know that we really need a special lib for it and we can take a look at what lodash does under the hood, as needed. If we were doing much more complex calculations (e.g. the greatest common denominator stuff we get from the math lib), I'd definitely be more worried about doing it by hand.

Reply:
The PR isn't 100% ready yet (hoping I'll give it one more pass in the morning and then un-draft it), but you can see an example of what I'm thinking hereabouts in the updated table legend values PR. https://github.com/perses/perses/pull/1237/files#r1227440821